orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/image/ImageTagger.java

```diff
@@ -18,32 +18,88 @@
 import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.netflix.spinnaker.orca.clouddriver.OortService;
 import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import org.slf4j.Logger;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 
-public interface ImageTagger {
-  String OPERATION = "upsertImageTags";
+import static java.lang.String.format;
+
+public abstract class ImageTagger {
+  protected static final String OPERATION = "upsertImageTags";
+
+  protected OortService oortService;
+  protected ObjectMapper objectMapper;
+  protected Logger log;
 
   /**
    * @return the OperationContext object that contains the cloud provider-specific list of operations as well as
    * cloud provider-specific output key/value pairs to be included in the task's output.
    */
-  OperationContext getOperationContext(Stage stage);
+  abstract protected OperationContext getOperationContext(Stage stage);
 
   /**
    * @return true when, according to the underlying cloud provider, image tags have been updated to match the respective target
    */
-  boolean areImagesTagged(Collection<Image> targetImages, Stage stage);
+  abstract protected boolean areImagesTagged(Collection<Image> targetImages, Stage stage);
 
   /**
    * @return The cloud provider type that this object supports.
    */
-  String getCloudProvider();
+  abstract protected String getCloudProvider();
+
+  protected ImageTagger(OortService oortService, ObjectMapper objectMapper, Logger log) {
+    this.oortService = oortService;
+    this.objectMapper = objectMapper;
+    this.log = log;
+  }
+
+  protected Collection findImages(Collection<String> imageNames, Stage stage, Class matchedImageType) {
+    if (imageNames == null || imageNames.isEmpty()) {
+      imageNames = new HashSet<>();
+
+      // attempt to find upstream images in the event that one was not explicitly provided
+      Collection<String> upstreamImageIds = ImageTaggerSupport.upstreamImageIds(stage, getCloudProvider());
+      if (upstreamImageIds.isEmpty()) {
+        throw new IllegalStateException("Unable to determine source image(s)");
+      }
+
+      for (String upstreamImageId : upstreamImageIds) {
+        // attempt to lookup the equivalent image name (given the upstream amiId/imageId)
+        List<Map> allMatchedImages = oortService.findImage(getCloudProvider(), upstreamImageId, null, null, null);
+        if (allMatchedImages.isEmpty()) {
+          throw new ImageNotFound(format("No image found (imageId: %s)", upstreamImageId), true);
+        }
+
+        String upstreamImageName = (String) allMatchedImages.get(0).get("imageName");
+        imageNames.add(upstreamImageName);
+
+        log.info(format("Found upstream image '%s' (executionId: %s)", upstreamImageName, stage.getExecution().getId()));
+      }
+    }
+
+    Collection foundImages = new ArrayList();
```

Comment:
Lost a fight here involving the type arg and java 8 type inferencing so had to drop the stream/map in favor of the older-style for-each.